### DISCUSSION TITLE
Atomic clock simulation

### DISCUSSION THREAD

**Discussion by nikkujensen** [aewallin/allantools#149] [Ideas]
Is it outside the scope to include functions that simulate atomic clock models? I work in the area of time generation and time synchronization and we often (1) simulate some clocks based on a dynamic model (2) compute the Allan variance and other statistics.

**Comment by aewallin** [comment]
Allantools already has some noise-generation functions for various types of phase/frequency noise.

By clock-models do you mean clock-ensemble algorithms using e.g. Kalman filters? Or something else?

If it's closely related to time & frequency statistics, and not a huge amount of code that is hard for me or someone else to review - then I think it could well fit into the allantools library. It helps if your models are based on published papers that we can reference. Check that you are OK with the LPGPL v3+ license.

**Reply by nikkujensen** [reply]
Thanks for the reply!

>By clock-models do you mean clock-ensemble algorithms using e.g. Kalman filters? Or something else?
By clock models, I mean something like these papers https://iopscience.iop.org/article/10.1088/0026-1394/45/6/S23 and https://iopscience.iop.org/article/10.1088/1681-7575/ad02fe. A discrete-time state space model is used to represent the composite noise of an atomic clock. Equations (69) from the paper shows the equation for simulating the 2nd order clock (with white frequency and white phase noise).

As for your comment about clock-ensembles, those will build on have a clock model that you can simulate. I am not thinking of introducing clock-ensemble algorithms here. Only the clock models that can be simulated

>If it's closely related to time & frequency statistics, and not a huge amount of code that is hard for me or someone else to review - then I think it could well fit into the allantools library.

It is certainly related to time & frequency topics. And the coding is short. One simple way to code it is by using some default functions in scipy like `scipy.signal.StateSpace`
